allennlp/modules/conditional_random_field.py

```diff
@@ -9,86 +9,135 @@
-                ])
-
-                if is_allowed:
-                    allowed.append((i, j))
-
-        # start transitions
-        for i, (to_bioul, *to_entity) in tokens.items():
-            if to_bioul in ('O', 'B', 'U'):
-                allowed.append((start_tag, i))
-
-        # end transitions
-        for i, (from_bioul, *from_entity) in tokens.items():
-            if from_bioul in ('O', 'L', 'U'):
-                allowed.append((i, end_tag))
-
-    elif constraint_type == "BIO":
-        for i, (from_bio, *from_entity) in tokens.items():
-            for j, (to_bio, *to_entity) in tokens.items():
+    for from_label_index, from_label in labels_with_boundaries:
+        if from_label in ("START", "END"):
+            from_tag = from_label
+            from_entity = ""
+        else:
+            from_tag = from_label[0]
+            from_entity = from_label[1:]
+        for to_label_index, to_label in labels_with_boundaries:
+            if to_label in ("START", "END"):
+                to_tag = to_label
+                to_entity = ""
+            else:
+                to_tag = to_label[0]
+                to_entity = to_label[1:]
+            if is_transition_allowed(constraint_type, from_tag, from_entity,
+                                     to_tag, to_entity):
+                allowed.append((from_label_index, to_label_index))
+    return allowed
 
-                is_allowed = any([
-                        # Can always transition to O or B-x
-                        to_bio in ('O', 'B'),
-                        # Can only transition to I-x from B-x or I-x
-                        to_bio == 'I' and from_bio in ('B', 'I') and from_entity == to_entity
-                ])
 
-                if is_allowed:
-                    allowed.append((i, j))
+def is_transition_allowed(constraint_type: str,
+                          from_tag: str,
+                          from_entity: str,
+                          to_tag: str,
+                          to_entity: str):
+    """
+    Given a constraint type and strings ``from_tag`` and ``to_tag`` that
+    represent the origin and destination of the transition, return whether
+    the transition is allowed under the given constraint type.
 
-        # start transitions
-        for i, (to_bio, *to_entity) in tokens.items():
-            if to_bio in ('O', 'B'):
-                allowed.append((start_tag, i))
+    Parameters
+    ----------
+    constraint_type : ``str``, required
+        Indicates which constraint to apply. Current choices are
+        "BIO", "IOB1", and BIOUL".
+    from_tag : ``str``, required
+        The tag that the transition originates from. For example, if the
+        label is ``I-PER``, the ``from_tag`` is ``I``.
+    from_entity: ``str``, required
+        The entity corresponding to the ``from_tag``. For example, if the
+        label is ``I-PER``, the ``from_entity`` is ``PER``.
```

Comment:
Technically, because you're using `from_label[1:]`, you'll actually get `-PER`, not `PER`.  But, that probably doesn't matter...

Reply:
Yeah, i opted to do this since the original behavior with `(tag, *entity)` meant `entity = ['-', 'P', 'E', 'R']`